### PR TITLE
[T5] Add clue finetune scripts of trainer

### PR DESCRIPTION
**Pull request by wjx-sudo**
### PR types

New features
### PR changes

APIs
### Description

新增t5_clue_trainer.py脚本，实现t5模型在clue的afqmc（语义相似度）、tnews（文本分类）、iflytek（长文本分类）、cmnli（自然语言推理）、ocnli（自然语言推理）、cluewsc2020（代词消歧）、csl（论文关键词识别）数据集上的fintune；
在data.py中增加CLUE_PROCESSED用于数据集预处理；
在units.py中增加各个clue数据集的模型评价指标。

### CODE REVIEW

**Comment by ZHUI** [comment]
PR types
PR changes
Description

这些补充一下，还有就是模型在各个数据集上，测试出来的效果。

File: examples/language_model/t5/data.py [hunk 1]
```diff
@@ -132,12 +276,22 @@ def get_train_dataloader(tokenizer, args):
...
+    def batchify_fn(
+        samples,
+        fn=Tuple(
+            Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # input_ids
+            Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # attention_mask
+            Pad(axis=0, pad_val=-100, dtype="int64"),  # lm_labels
+            Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # decoder_attention_mask
+        ),
+    ):
+        return fn(samples)
```
Comment (ZHUI) [inline examples/language_model/t5/data.py:294]: 这里只是为了过 pre-commit 的修改吧

File: examples/language_model/t5/run_clue_trainer.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+            "help": "Path to pretrained model or model identifier from https://paddlenlp.readthedocs.io/zh/latest/model_zoo/transformers.html"
+        },
+    )
+    export_model_dir: Optional[str] = field(
+        default=None,
+        metadata={"help": "Path to directory to store the exported inference model."},
+    )
+
+
+class T5ClueTrainer(Trainer):
```
Comment (ZHUI) [inline examples/language_model/t5/run_clue_trainer.py]: https://github.com/huggingface/transformers/blob/7119bb052a3f492b9af3afe4f3f13132445eba6e/examples/pytorch/translation/run_translation.py#L560-L568

这块看 能不能 使用 Seq2SeqTrainer， paddlenlp 中已经有了，但比较少例子

File: examples/language_model/t5/run_clue_trainer.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+        samples,
+        fn=BatchDict(
+            {
+                "input_ids": Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # input_ids
+                "attention_mask": Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # attention_mask
+                "labels": Pad(axis=0, pad_val=-100, dtype="int64"),  # lm_labels
+                "decoder_attention_mask": Pad(
+                    axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"
+                ),  # decoder_attention_mask
+            }
```
Comment (ZHUI) [inline examples/language_model/t5/run_clue_trainer.py]: 这里 也看 换成 DataCollatorForSeq2Seq

https://github.com/huggingface/transformers/blob/7119bb052a3f492b9af3afe4f3f13132445eba6e/src/transformers/data/data_collator.py#L518-L605

File: examples/language_model/t5/run_clue_trainer.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    else:
+        ds = load_dataset("clue", args.task_name, splits=split)
+        ds.map(
+            partial(trans_func, tokenizer=tokenizer, args=args),
+            batched=False,
+            lazy=False,
+        )
+        save_pickle(ds, filename)
+
+    return ds
```
Comment (ZHUI) [inline examples/language_model/t5/run_clue_trainer.py]: 可以删除吧

File: examples/language_model/t5/run_clue_trainer.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+                    pred = self.label2id[pred]
+            else:
+                # for regression task.
+                label = float(label.replace(" ", ""))
+                try:
+                    pred = float(pred.replace(" ", ""))
+                except Exception as e:
+                    # set to zero if the generated text can not convert to float
+                    pred = 0.0
+                    print(e)
```
Comment (ZHUI) [inline examples/language_model/t5/run_clue_trainer.py]: 后处理

Comment (ZHUI) [reply examples/language_model/t5/run_clue_trainer.py]: https://github.com/huggingface/transformers/blob/7119bb052a3f492b9af3afe4f3f13132445eba6e/examples/pytorch/translation/run_translation.py#L538-L557

放到 metric 计算中

File: paddlenlp/trainer/trainer_seq2seq.py [hunk 6]
```diff
@@ -203,12 +209,15 @@ def prediction_step(
 
         with paddle.no_grad():
             if has_labels:
-                with self.compute_loss_context_manager():
-                    outputs = model(**inputs)
-                if self.label_smoother is not None:
-                    loss = self.label_smoother(outputs, inputs["labels"]).mean().detach()
+                if hasattr(self, "compute_loss_context_manager"):
+                    with self.compute_loss_context_manager():
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer_seq2seq.py]: 这里应该 hf 重构过。

Comment (ZHUI) [reply paddlenlp/trainer/trainer_seq2seq.py]: https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2537-L2541

Comment (ZHUI) [reply paddlenlp/trainer/trainer_seq2seq.py]: autocast_smart_context_manager

File: paddlenlp/trainer/trainer_seq2seq.py [hunk 7]
```diff
@@ -193,6 +197,8 @@ def prediction_step(
             generation_inputs,
             **gen_kwargs,
         )
+        if isinstance(generated_tokens, tuple):
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer_seq2seq.py:197]: 说明：输出与hf不一致，多了xxx

File: paddlenlp/trainer/trainer_seq2seq.py [hunk 8]
```diff
@@ -184,7 +184,11 @@ def prediction_step(
         # prepare generation inputs
         # some encoder-decoder models can have varying encoder's and thus
         # varying model input names
-        if hasattr(self.model, "encoder") and self.model.encoder.main_input_name != self.model.main_input_name:
```
Comment (ZHUI) [inline paddlenlp/trainer/trainer_seq2seq.py]: main_input_name 模型里面加。

Comment (ZHUI) [reply paddlenlp/trainer/trainer_seq2seq.py]: https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L982-L1014

Comment (ZHUI) [reply paddlenlp/trainer/trainer_seq2seq.py]: model_utils.py

File: examples/language_model/t5/run_clue_trainer.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+        return results
+
+    trainer = Seq2SeqTrainer(
+        model=model,
+        args=training_args,
+        train_dataset=train_dataset if training_args.do_train else None,
+        eval_dataset=eval_dataset if training_args.do_eval else None,
+        tokenizer=tokenizer,
+        data_collator=data_collator,
+        compute_metrics=compute_metrics if training_args.predict_with_generate else None,
```
Comment (ZHUI) [inline examples/language_model/t5/run_clue_trainer.py]: fix it

File: paddlenlp/transformers/t5/modeling.py [hunk 10]
```diff
@@ -824,6 +824,8 @@ def _shift_right(self, input_ids):
 
 
 class T5Stack(nn.Layer):
+    main_input_name = "input_ids"
```
Comment (wjx-sudo) [inline paddlenlp/transformers/t5/modeling.py]: 在t5 encoder部分增加属性main_input_name

File: paddlenlp/trainer/trainer_seq2seq.py [hunk 11]
```diff
@@ -197,6 +193,7 @@ def prediction_step(
             generation_inputs,
             **gen_kwargs,
         )
+        # different from hf returns: tuple[Tensor]: It is a tuple contains two elements: ids and scores.
```
Comment (wjx-sudo) [inline paddlenlp/trainer/trainer_seq2seq.py:196]: 新增注释说明

File: paddlenlp/transformers/t5/modeling.py [hunk 12]
```diff
@@ -824,6 +824,8 @@ def _shift_right(self, input_ids):
 
 
 class T5Stack(nn.Layer):
```
Comment (ZHUI) [inline paddlenlp/transformers/t5/modeling.py]: https://github.com/huggingface/transformers/blob/main/src/transformers/models/t5/modeling_t5.py#L853 

这里没有继承 `main_input_name` 是因为基类不对。

要继承 `T5PretrainedModel `

File: paddlenlp/transformers/t5/modeling.py [hunk 13]
```diff
@@ -823,11 +823,9 @@ def _shift_right(self, input_ids):
         return shifted_input_ids
 
 
-class T5Stack(nn.Layer):
-    main_input_name = "input_ids"
-
+class T5Stack(T5PretrainedModel):
     def __init__(self, config: T5Config, embed_tokens: Optional[nn.Embedding] = None):
-        super().__init__()
+        super().__init__(config)
```
Comment (wjx-sudo) [inline paddlenlp/transformers/t5/modeling.py:828]: 修改继承基类T5PretrainedModel

File: paddlenlp/trainer/trainer.py [hunk 14]
```diff
@@ -257,7 +257,11 @@ def __init__(
         self.compute_metrics = compute_metrics
         self.preprocess_logits_for_metrics = preprocess_logits_for_metrics
         self.optimizer, self.lr_scheduler = optimizers
-
```
Comment (wjx-sudo) [inline paddlenlp/trainer/trainer.py:260]: hf中在trainer_pt_utils新增 LabelSmoother
https://github.com/huggingface/transformers/blob/main/src/transformers/trainer_pt_utils.py#L472
暂时没有在paddle trainer_utils中新增

Comment (ZHUI) [reply paddlenlp/trainer/trainer.py:260]: 赞，这个咱们先不支持了。后续有需求再考虑支持

File: examples/language_model/t5/data.py [hunk 15]
```diff
@@ -166,12 +320,22 @@ def get_dev_dataloader(tokenizer, args):
 
     batch_sampler = BatchSampler(ds, batch_size=args.train_batch_size, shuffle=False)
 
-    batchify_fn = lambda samples, fn=Tuple(
-        Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # input_ids
-        Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # attention_mask
-        Pad(axis=0, pad_val=-100, dtype="int64"),  # lm_labels
-        Pad(axis=0, pad_val=tokenizer.pad_token_id, dtype="int64"),  # decoder_attention_mask
-    ): fn(samples)
+    # batchify_fn = lambda samples, fn=Tuple(
```
Comment (ZHUI) [inline examples/language_model/t5/data.py]: 删除

File: examples/language_model/t5/data.py [hunk 16]
```diff
@@ -203,12 +367,22 @@ def get_mnli_dev_dataloader(tokenizer, args, matched=True):
...
```
Comment (ZHUI) [inline examples/language_model/t5/data.py]: 删除

File: examples/language_model/t5/run_clue_trainer.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+# limitations under the License.
+
+import os
+from dataclasses import dataclass, field
+from functools import partial
+from typing import Optional
+
+import paddle
+
+# import paddle.nn as nn
```
Comment (ZHUI) [inline examples/language_model/t5/run_clue_trainer.py]: 删除

**Review by ZHUI — DISMISSED** [review]
差不多了，把一些 没用的注释删掉就好了

**Review by ZHUI — APPROVED** [review]
LGTM

**Comment by ZHUI** [comment]
@LemonNoel 后续改造 verbalizer 支持生成类，有监督问题
